lib/types.nix

```diff
@@ -27,6 +27,13 @@ rec {
 
   maybeRaw = type: types.either type rawLua;
 
+  booleanInt =
+    with types;
+    coercedTo bool (v: if v then 1 else 0) (enum [
```

Comment:
Do we want to use `coercedTo` (both types officially supported and documented, no warning shown) or `transitionType` (one type is deprecated and shows a warning)?

Reply:
I think we need a `coercedTo` or a `transitionType`, even if long-term we'd rather have unmutated `settings` configs.

Reply:
To summarize, we're choosing between:

```nix
# Option 1 - no coercion
boolInt = enum [ 0 1 ];
# Option 2 - silent coercion
intBool = coercedTo bool (v: if v then 1 else 0) (enum [ 0 1 ]);
# Option 3 - deprecated coercion
booleanInt = transitionType bool (v: if v then 1 else 0) (enum [ 0 1 ]);
```

Currently, existing configs are silently broken.
- If we go with option 1, existing configs will get eval errors\*.
- If we go with option 2, existing configs will start working and we will be supporting a coerced type, like we already do with `strLua`.
- If we go with option 3, existing configs will start working, but will get warning messages\*\* telling them to convert to `enum 0 or 1` definitions.

<details><summary>Details</summary>
<p>

\*The actual error printed by option 1:

```
error: A definition for option `plugins.autosource.settings.prompt_for_changed_file' is not of type `null or one of 0, 1 or raw lua code'. Definition values:
- In `<unknown-file>': true
```

\*\*Example warning printed by option 3:

```
Passing a boolean for `plugins.autosource.settings.prompt_for_changed_file' is deprecated, use one of 0, 1 instead. Definitions: 
- In `<unknown-file>': true
```

</p>
</details>

Option 2 is a similar approach as used by `strLua`, however we plan to transition that to option 3 after 24.11. See https://github.com/nix-community/nixvim/issues/2003

As discussed the other day on matrix, the consensus is option 2 is the worst option because we are silently fixing something without users noticing, _and_ we are continuing to advertise `bool` support in docs when we plan to later deprecate it...

Therefore I think option 3 is the best approach. The warnings will highlight that something has changed, and will draw attention to options configured that previously had no effect.

It may also motivate users to change their config to use the plugin's expected value without us needing to coerce it.